opamp-client/src/main/java/io/opentelemetry/opamp/client/internal/OpampClient.java

```diff
@@ -0,0 +1,97 @@
+   *
+   * @param remoteConfigStatus The new remote config status.
+   */
+  void setRemoteConfigStatus(Opamp.RemoteConfigStatus remoteConfigStatus);
+
+  /**
+   * Sets the current effective config which will be sent in the next agent to server request.
+   *
+   * @param effectiveConfig The new effective config.
+   */
+  void setEffectiveConfig(Opamp.EffectiveConfig effectiveConfig);
+
+  interface Callback {
+    /**
+     * Called when the connection is successfully established to the Server. May be called after
+     * {@link #start(OpampClient.Callback)} is called and every time a connection is established to
+     * the Server. For WebSocket clients this is called after the handshake is completed without any
+     * error. For HTTP clients this is called for any request if the response status is OK.
+     *
+     * @param client The relevant {@link OpampClient} instance.
+     */
+    void onConnect(OpampClient client);
+
+    /**
+     * Called when the connection to the Server cannot be established. May be called after {@link
+     * #start(OpampClient.Callback)} is called and tries to connect to the Server. May also be
+     * called if the connection is lost and reconnection attempt fails.
+     *
+     * @param client The relevant {@link OpampClient} instance.
+     * @param throwable The exception.
+     */
+    void onConnectFailed(OpampClient client, Throwable throwable);
+
+    /**
+     * Called when the Server reports an error in response to some previously sent request. Useful
+     * for logging purposes. The Agent should not attempt to process the error by reconnecting or
+     * retrying previous operations. The client handles the ErrorResponse_UNAVAILABLE case
+     * internally by performing retries as necessary.
+     *
+     * @param client The relevant {@link OpampClient} instance.
+     * @param errorResponse The error returned by the Server.
+     */
+    void onErrorResponse(OpampClient client, Opamp.ServerErrorResponse errorResponse);
+
+    /**
+     * Called when the Agent receives a message that needs processing. See {@link MessageData}
+     * definition for the data that may be available for processing. During onMessage execution the
+     * {@link OpampClient} functions that change the status of the client may be called, e.g. if
+     * RemoteConfig is processed then {@link
+     * #setRemoteConfigStatus(opamp.proto.Opamp.RemoteConfigStatus)} should be called to reflect the
+     * processing result. These functions may also be called after onMessage returns. This is
+     * advisable if processing can take a long time. In that case returning quickly is preferable to
+     * avoid blocking the {@link OpampClient}.
+     *
+     * @param client The relevant {@link OpampClient} instance.
+     * @param messageData The server response data that needs processing.
+     */
+    void onMessage(OpampClient client, MessageData messageData);
+  }
+}
```

Comment:
I assume more callbacks will be added later, e.g. [OnOpampConnectionSettings](https://pkg.go.dev/github.com/open-telemetry/opamp-go@v0.16.0/client/types#CallbacksStruct.OnOpampConnectionSettings), OnCommand, etc.

Reply:
Correct, I think those should be added as soon as we add support for those functionalities (unless we see a reason to add them beforehand with noop functionality).